lib/std/ring.zig

```diff
@@ -0,0 +1,304 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const Atomic = std.atomic.Atomic;
+const expectEqual = std.testing.expectEqual;
+
+pub fn Ring(comptime T: type, comptime size: usize) type {
+    return if (builtin.single_threaded)
+        RingGeneric(T, size, .non_atomic)
+    else
+        RingGeneric(T, size, .mpmc);
+}
+
+const RingKind = enum {
+    mpmc,
+    non_atomic,
+};
+
+/// A circular buffer with a tail and a head. The tail is an index within the
+/// buffer from which the next item will be read. The head is an index within
+/// the buffer to which the next item will be written. Both the tail and
+/// the head are atomically incremented via a CAS loop and are bound to the
+/// range of `0..std.math.maxInt(usize)`. The size is comptime asserted to be
+/// a power of two. This permits the tail and head to overflow while remaining
+/// aligned to the size. Every slot within the buffer is tagged with an atomic
+/// enum indicating the current state of the slot (occupied or empty).
+///
+/// *Credit*: This draws upon the Linux Kernel's BPF ring buffer specification
```

Comment:
The document is phrased "This document describes BPF ring buffer design, API, and implementation details." and it is more like a reference manual rather than limited "user manual/spec".

Reply:
Frankly, I fail to see the difference between a "reference manual" and a "user manual/spec." Would "documentation" suffice instead of "specification?"

Reply:
yes.